client/tokenized-payment-request/event-handlers.js

```diff
@@ -0,0 +1,157 @@
+				status: 'invalid_shipping_address',
+			} );
+
+			return;
+		}
+
+		event.updateWith( {
+			// Possible statuses: https://docs.stripe.com/js/appendix/payment_response#payment_response_object-complete
+			status: 'success',
+			shippingOptions: transformCartDataForShippingOptions( cartData ),
+			total: {
+				label: getPaymentRequestData( 'total_label' ),
+				amount: transformPrice(
+					parseInt( cartData.totals.total_price, 10 ) -
+						parseInt( cartData.totals.total_refund || 0, 10 ),
+					cartData.totals
+				),
+			},
+			displayItems: transformCartDataForDisplayItems( cartData ),
+		} );
+	} catch ( error ) {
+		// Possible statuses: https://docs.stripe.com/js/appendix/payment_response#payment_response_object-complete
+		event.updateWith( {
+			status: 'fail',
+		} );
+	}
+};
+
+export const shippingOptionChangeHandler = async ( event ) => {
+	try {
+		const cartData = await cartApi.selectShippingRate( {
+			package_id: 0,
+			rate_id: event.shippingOption.id,
+		} );
+
+		event.updateWith( {
+			status: 'success',
+			total: {
+				label: getPaymentRequestData( 'total_label' ),
+				amount: transformPrice(
+					parseInt( cartData.totals.total_price, 10 ) -
+						parseInt( cartData.totals.total_refund || 0, 10 ),
+					cartData.totals
+				),
+			},
+			displayItems: transformCartDataForDisplayItems( cartData ),
+		} );
+	} catch ( error ) {
+		event.updateWith( { status: 'fail' } );
+	}
+};
+
+const paymentResponseHandler = async (
+	api,
+	response,
+	completePayment,
+	abortPayment,
+	event
+) => {
+	if ( response.payment_result.payment_status !== 'success' ) {
+		return abortPayment(
+			event,
+			getErrorMessageFromNotice( response.message )
+		);
+	}
+
+	try {
+		const confirmationRequest = api.confirmIntent(
+			response.payment_result.redirect_url
+		);
+		// We need to call `complete` outside of `completePayment` to close the dialog for 3DS.
+		event.complete( 'success' );
+
+		// `true` means there is no intent to confirm.
+		if ( confirmationRequest === true ) {
+			completePayment( response.payment_result.redirect_url );
+		} else {
+			const redirectUrl = await confirmationRequest;
+
+			completePayment( redirectUrl );
+		}
+	} catch ( error ) {
+		abortPayment( event, error.message );
+	}
+};
+
+export const paymentMethodHandler = async (
+	api,
+	completePayment,
+	abortPayment,
+	event
+) => {
+	try {
+		// Kick off checkout processing step.
+		const response = await cartApi.placeOrder( {
+			// adding extension data as a separate action,
+			// so that we make it harder for external plugins to modify or intercept checkout data.
+			...transformStripePaymentMethodForStoreApi( event ),
+			extensions: applyFilters(
+				'wcpay.payment-request.cart-place-order-extension-data',
+				{}
+			),
+		} );
+
+		paymentResponseHandler(
+			api,
+			response,
+			completePayment,
+			abortPayment,
+			event
+		);
+	} catch ( error ) {
+		abortPayment( event, error.message );
```

Comment:
This isn't really a big deal and honestly it's probably for the best that we don't necessarily output these error messages that could be generated by Stripe or other sources directly to the client, but FWIW, when I manually broke the checkout by throwing an exception, this `error` return _didn't_ have a `message` property. This is what it looks like for me.

![Nested error message](https://cdn-std.droplr.net/files/acc_1104206/8m0NDz)
_Actual error message nested inside `payment_result`_

Did you notice this as well or should we expect that a message should be present in this property?

Nonetheless, I think this is still mostly fine, since I believe we are logging these to the back-end anyways and we are still handling the exception and returning an unexplained error, which is probably good enough. 🤷

Reply:
I was seeing the error with a `message` property but I see that it depends on from where the error is thrown. Thanks @frosso for sharing how you handled both formats!

I noticed that the errors could include HTML so I've wrapped them with `getErrorMessageFromNotice()` similar to other errors.